src/docs/sphinx/dev_guide/state_manager.rst

```diff
@@ -0,0 +1,125 @@
+
+``MFEMSidreDataCollection`` is an implementation of
+MFEM's `DataCollection interface <https://mfem.github.io/doxygen/html/classmfem_1_1DataCollection.html>`_,
+which allows for instances of ``mfem::GridFunction`` and ``mfem::QuadratureFunction`` to be associated
+with a single instance of ``mfem::Mesh``.  Because Serac supports multiple meshes within a given simulation,
+``StateManager`` acts as an abstraction over multiple ``MFEMSidreDataCollection`` instances.
+
+
+Nominal (non-restart) Workflow
+------------------------------
+
+The first interaction a user must always make with ``StateManager`` is a call to its static
+``initialize()`` method.  ``StateManager`` is implemented as a global singleton so that its
+contained data can be accessed from physics modules anywhere in the simulation.  The singleton
+is initialized with a non-owning reference to an ``axom::sidre::DataStore`` and an output directory
+to which data will be saved.
+
+.. note:: ``StateManager`` does not own its ``DataStore`` because Serac uses a single datastore to
+  store different kinds of data - that is, data unrelated to the state defined above.  In particular,
+  input file data is also stored in the per-simulation ``DataStore`` instance.
+
+Before any other kinds of state can be created, a mesh must be registered via ``SetMesh()``.
+In order for a restart to work properly, all state data must be owned by the underlying
+``StateManager``, so ownership of the mesh is transferred via a ``unique_ptr``.  In the case of multi-mesh
+simulations, a name should also be used to uniquely identify the mesh.
+
+Individual physics modules - that are of course based on these kinds of state - can now be constructed.
+In general, this process looks something like the following:
+
+  1. The physics module constructor accepts a mesh pointer and forwards it to the ``BasePhysics``
+     constructor. This parameter is required only in multi-mesh configurations and defaults to the
+     default mesh otherwise.  Although meshes also have string-valued tags associated with them,
+     a user of a physics module would find it more intuitive to pass the mesh pointer they wish to use.
+     Specifically, the meaning of a mesh parameter is much easier to discern than a string parameter.
+
+  #. The physics module creates its fields (e.g., temperature for a thermal conduction module) via
+     calls to ``StateManager::newState()``.  In addition to the ``serac::FiniteElementState`` constructor
+     arguments, this method also accepts a string-valued tag for the mesh with which the field is
+     associated.  The appropriate tag is a member of ``BasePhysics`` and initialized in the previous step.
+     FIXME: Should we provide a protected helper method in ``BasePhysics`` so derived modules don't need
+     to reference the member explicitly? Or perhaps ``StateManager::newState()`` et al should just take
+     a mesh pointer instead of a tag?
```

Comment:
How can it take a mesh pointer? If I recall correctly, `mfem::Mesh` does not have an equality check.

Reply:
The current logic uses raw pointer comparisons, anticipating the use case where the user does something like:
```cpp
Mesh m = /* ... */;
StateManager::setMesh(m, "primary");
// ... pointers would be the same
ThermalConduction tc(..., m);
Solid s(..., m);

Reply:
I think this is a fair assumption.